jdaviz/configs/specviz/plugins/viewers.py

```diff
@@ -57,21 +57,34 @@ def __init__(self, *args, **kwargs):
         self.figure.marks = self.figure.marks + self._offscreen_lines_marks.marks
 
         self.state.add_callback('show_uncertainty', self._show_uncertainty_changed)
+
+        self._expected_subset_layers = []
+        self.state.add_callback('layers', self._on_layers_update)
         self.display_mask = False
 
         # Change collapse function to sum
         self.state.function = 'sum'
 
-    def _on_subset_create(self, msg):
+    def _on_layers_update(self, *args):
+        if not len(self._expected_subset_layers):
+            return
+        # we'll make a deepcopy so that we can remove entries from the self._expected_subset_layers
+        # to avoid recursion, but also handle multiple layers for the same subset
+        expected_subset_layers = self._expected_subset_layers[:]
         for layer in self.state.layers:
-            if layer.layer.label == msg.subset.label:
-                layer.linewidth = 3
+            if layer.layer.label in expected_subset_layers:
+                if layer.layer.label in self._expected_subset_layers:
+                    self._expected_subset_layers.remove(layer.layer.label)
+                if isinstance(layer.layer.subset_state, RoiSubsetState):
+                    layer.linewidth = 1
+                else:
+                    layer.linewidth = 3
 
-    def _on_add_data(self, msg):
-        data_label = msg.data.label
-        for layer in self.state.layers:
-            if "Subset" in layer.layer.label and layer.layer.data.label == data_label:
-                layer.linewidth = 3
+    def _on_subset_create(self, msg):
+        # we don't have access to the actual subset yet to tell if its spectral or spatial, so
+        # we'll store the name of this new subset and change the default linewidth when the
+        # layers are added
+        self._expected_subset_layers.append(msg.subset.label)
```

Comment:
I'm not exactly following what `self._expected_subset_layers` is doing here. It appears to me that this list is being populated whenever subsets are created and then checked against in `_on_layers_update`. I'm not seeing how this list ever deviates from the subset drawer contents, in which case I would advocate for grabbing this list directly from the subset drawer. But I assume I'm missing something here

Reply:
Let me try to explain the logic, and then please give any suggestions for clearer names or inline-comments (I struggled with the name myself).

Every time a subset is added, `_on_subset_create` is triggered, but the `msg.subset` object is essentially empty (it hasn't yet been processed fully by glue).  The code to set the default linewidth to 3 used to live here, but I couldn't not figure out any way to tell whether it was spectral or spatial.  After this is triggered, glue creates the subset layer objects, triggering `_on_layers_update` which have access to `layer.layer.subset_state` which can now tell the difference between spectral and spatial.  We can't move all the logic there though, since then any change to a subset would _revert_ user-applied linewidth choices.  We only want to apply the default on a layer that is updated by the addition of a new subset.  So, when the subset is created, I append the name of the subset to `self._expected_subset_layers`.  That list essentially means ("subsets that have been added and are expecting new layers to be created, at which point we can set defaults").  When the layer is then added, we can know to only act on subsets with those labels and remove them from the list so that they don't revert to the default from future changes.

Reply:
Ok I think this is starting to make sense. That being said, I'm struggling to find better names here. I feel like in-line comments will probably do us better here

But before I start making suggestions, I'm surprised there isn't a better trigger to use here. Like something of the nature of `_on_layer_create`? That would make this so much easier to navigate